agent/system_metrics.go

```diff
@@ -0,0 +1,208 @@
+	// Collect returns the value of the collected metric when the function is called.
+	Collect() ([]byte, error)
+}
+
+type SystemMetricCollector struct {
+	mu          sync.Mutex
+	wg          sync.WaitGroup
+	innerCancel context.CancelFunc
+	logger      client.LoggerProducer
+	taskOpts    *system_metrics.SystemMetricsOptions
+	connOpts    system_metrics.ConnectionOptions
+	interval    time.Duration
+	collectors  []MetricCollector
+	id          string
+	client      *system_metrics.SystemMetricsClient
+	errored     bool
+	closed      bool
+}
+
+type DiskUsageCollector struct{}
+
+type UptimeCollector struct{}
+
+type ProcessCollector struct{}
+
+// DataFormat describes how the time series data is stored.
+type DataFormat int32
+
+// Valid DataFormat values.
+const (
+	DataFormatText DataFormat = 0
+	DataFormatFTDC DataFormat = 1
+	DataFormatBSON DataFormat = 2
+	DataFormatJSON DataFormat = 3
+	DataFormatCSV  DataFormat = 4
+)
+
+type UptimeWrapper struct {
+	Uptime uint64 `json:"uptime"`
+}
+
+type ProcessesWrapper struct {
+	Processes []ProcessWrapper
+}
+
+type ProcessWrapper struct {
+	Pid           int32          `json:"pid"`
+	CPUPercent    float64        `json:"%cpu"`
+	MemoryPercent float32        `json:"%mem"`
+	VSZ           uint64         `json:"vsz"`
+	RSS           uint64         `json:"rss"`
+	TT            string         `json:"tt"`
+	Stat          string         `json:"stat"`
+	Started       int64          `json:"started"`
+	Time          *cpu.TimesStat `json:"time"`
+	Command       string         `json:"command"`
+}
+
+func (collector *DiskUsageCollector) Name() string {
+	return "disk_usage"
+}
+
+func (collector *UptimeCollector) Name() string {
+	return "uptime"
+}
+
+func (collector *ProcessCollector) Name() string {
+	return "process"
+}
+
+func (collector *DiskUsageCollector) Format() DataFormat {
+	return DataFormatFTDC
+}
+
+func (collector *UptimeCollector) Format() DataFormat {
+	return DataFormatFTDC
+}
+
+func (collector *ProcessCollector) Format() DataFormat {
+	return DataFormatJSON
+}
+
+func (collector *DiskUsageCollector) Collect(ctx context.Context) ([]byte, error) {
+	usage, err := disk.Usage("/")
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	return convertJSONToFTDC(ctx, usage)
+}
+
+func (collector *UptimeCollector) Collect(ctx context.Context) ([]byte, error) {
+	uptime, err := host.Uptime()
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	uptimeWrapper := UptimeWrapper{uptime}
+	return convertJSONToFTDC(ctx, uptimeWrapper)
+}
+
+func (collector *ProcessCollector) Collect(ctx context.Context) ([]byte, error) {
+	processes, err := process.Processes()
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	processesPopulated := make([]ProcessWrapper, len(processes))
+
+	for i, process := range processes {
+		cpuPercent, err := process.CPUPercent()
```

Comment:
should we be tracking these errors?

Reply:
A lot of them fail, I think for reasons beyond our control (e.g. the process ended in the middle of requesting the information). I don't think we need to track the errors